File: src/main/java/com/antlr/plugin/psi/ModeSpecNode.java
package com.antlr.plugin.psi;

import com.antlr.plugin.parser.PsiElementFactory;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.tree.IElementType;
import com.antlr.plugin.ANTLRv4TokenTypes;
import com.antlr.plugin.parser.ANTLRv4Lexer;
import com.antlr.plugin.parser.ANTLRv4Parser;
import org.jetbrains.annotations.NotNull;

import static com.antlr.plugin.psi.MyPsiUtils.findFirstChildOfType;


/**
 * A node representing a lexical {@code mode} definition, and all its child rules.
 *
 * @implNote this is technically not a 'rule', but it has the same characteristics as a named rule so we
 * can extend {@code RuleSpecNode}
 */
public class ModeSpecNode extends RuleSpecNode {

    public ModeSpecNode(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public IElementType getRuleRefType() {
        return ANTLRv4TokenTypes.TOKEN_ELEMENT_TYPES.get(ANTLRv4Lexer.TOKEN_REF);
    }

    @Override
    public GrammarElementRefNode getNameIdentifier() {
        PsiElement idNode = findFirstChildOfType(this, ANTLRv4TokenTypes.getRuleElementType(ANTLRv4Parser.RULE_identifier));

        if (idNode != null) {
            PsiElement firstChild = idNode.getFirstChild();

            if (firstChild instanceof GrammarElementRefNode) {
                return (GrammarElementRefNode) firstChild;
            }
        }

        return null;
    }

    public static class Factory implements PsiElementFactory {
        public static Factory INSTANCE = new Factory();

        @Override
        public PsiElement createElement(ASTNode node) {
            return new ModeSpecNode(node);
        }
    }
}

File: src/main/java/com/antlr/language/psi/ModeSpecNode.kt
package com.antlr.language.psi

import com.antlr.language.ANTLRv4Lexer
import com.antlr.language.ANTLRv4Parser
import com.antlr.language.AntlrTokenTypes
import com.antlr.language.psi.MyPsiUtils.findFirstChildOfType
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.tree.IElementType

/**
 * A node representing a lexical `mode` definition, and all its child rules.
 *
 * @implNote this is technically not a 'rule', but it has the same characteristics as a named rule so we
 * can extend `RuleSpecNode`
 */
class ModeSpecNode(node: ASTNode) : RuleSpecNode(node) {
    override val ruleRefType: IElementType
        get() = AntlrTokenTypes.TOKEN_ELEMENT_TYPES[ANTLRv4Lexer.TOKEN_REF]

    override fun getNameIdentifier(): GrammarElementRefNode? {
        val idNode: PsiElement? =
            findFirstChildOfType(this, AntlrTokenTypes.getRuleElementType(ANTLRv4Parser.RULE_identifier))

        if (idNode != null) {
            val firstChild = idNode.firstChild

            if (firstChild is GrammarElementRefNode) {
                return firstChild
            }
        }

        return null
    }

    class Factory : PsiElementFactory {
        override fun createElement(node: ASTNode): PsiElement {
            return ModeSpecNode(node)
        }

        companion object {
            var INSTANCE: Factory = Factory()
        }
    }
}
